### PR TITLE
[Feature] Support write quorum property for table

### LINKED ISSUE #11613: Support write quorum property for table
## Feature request
User can set `write_quorum` to specify replica write strategy when create/alter table.

**MAJORITY**: write majority replica successfully returns success, default option
**ONE**: write one replica successfully returns success
**ALL**: write all replica successfully returns success

> Note: The write will only return after all replicas have determined the state, not if the write to some replicas is successful and the state of other replicas is unknown

### CREATE TABLE
```sql
CREATE TABLE my_table
(
    k1 TINYINT,
    k2 DECIMAL(10, 2) DEFAULT "10.5",
    v1 CHAR(10) REPLACE,
    v2 INT SUM
)
ENGINE = olap
AGGREGATE KEY(k1, k2)
DISTRIBUTED BY HASH(k1) BUCKETS 10
PROPERTIES ("write_quorum" = "one");
```
### ALTER TABLE
```sql
ALTER TABLE my_table
SET ("write_quorum" = "ALL");
```

### PR DESCRIPTION
**Pull request by meegoo**
1. MAJORITY: write majority replica successfully returns success, default option
2. ONE: write one replica successfully returns success
3. ALL: write all replica successfully returns success

## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #11613

## Problem Summary(Required) ：

User can set `write_quorum` to specify replica write strategy when create/alter table.

**MAJORITY**: write majority replica successfully returns success, default option
**ONE**: write one replica successfully returns success
**ALL**: write all replica successfully returns success

> Note: The write will only return after all replicas have determined the state, not if the write to some replicas is successful and the state of other replicas is unknown

### CREATE TABLE
```sql
CREATE TABLE my_table
(
    k1 TINYINT,
    k2 DECIMAL(10, 2) DEFAULT "10.5",
    v1 CHAR(10) REPLACE,
    v2 INT SUM
)
ENGINE = olap
AGGREGATE KEY(k1, k2)
DISTRIBUTED BY HASH(k1) BUCKETS 10
PROPERTIES ("write_quorum" = "one");
```
### ALTER TABLE
```sql
ALTER TABLE my_table
SET ("write_quorum" = "ALL");
```

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: gensrc/proto/internal_service.proto [hunk 1]
```diff
@@ -149,6 +149,7 @@ message PTabletWriterOpenRequest {
     optional string txn_trace_parent = 22;
     optional bool is_replicated_storage = 23;
     optional int64 timeout_ms = 24;
+    optional InsertQuorumTypePB insert_quorum = 25;
```
Comment (howrocks) [inline gensrc/proto/internal_service.proto:152]: write_quorum is clearer? update、delete is also included in quorum.

File: fe/fe-core/src/main/java/com/starrocks/load/loadv2/BulkLoadJob.java [hunk 2]
```diff
@@ -99,6 +99,7 @@ public BulkLoadJob(long dbId, String label, OriginStatement originStmt) throws M
             sessionVariables.put(SessionVariable.SQL_MODE, Long.toString(var.getSqlMode()));
             sessionVariables.put(SessionVariable.LOAD_TRANSMISSION_COMPRESSION_TYPE, var.getloadTransmissionCompressionType());
             sessionVariables.put(SessionVariable.ENABLE_REPLICATED_STORAGE, Boolean.toString(var.getEnableReplicatedStorage()));
+            sessionVariables.put(SessionVariable.INSERT_QUORUM, var.getInsertQuorum());
```
Comment (howrocks) [inline fe/fe-core/src/main/java/com/starrocks/load/loadv2/BulkLoadJob.java:102]: Maybe we need a way to persistent this config? global level or per table level.

File: be/src/storage/segment_replicate_executor.cpp [hunk 3]
```diff
@@ -193,8 +193,13 @@ ReplicateToken::ReplicateToken(std::unique_ptr<ThreadPoolToken> replicate_pool_t
         _replicate_channels.emplace_back(std::move(std::make_unique<ReplicateChannel>(
                 opt, opt->replicas[i].host(), opt->replicas[i].port(), opt->replicas[i].node_id())));
     }
-    // default quorom policy
-    _max_fail_replica_num = opt->replicas.size() - (opt->replicas.size() / 2 + 1);
+    if (opt->write_quorum == WriteQuorumTypePB::ONE) {
+        _max_fail_replica_num = opt->replicas.size();
```
Comment (howrocks) [inline be/src/storage/segment_replicate_executor.cpp:197]: opt->replicas.size() - 1？

File: fe/fe-core/src/main/java/com/starrocks/load/loadv2/LoadingTaskPlanner.java [hunk 4]
```diff
@@ -100,10 +101,12 @@ public class LoadingTaskPlanner {
     private TupleDescriptor tupleDesc;
     private List<Pair<Integer, ColumnDict>> globalDicts = Lists.newArrayList();
 
+    private Map<String, String> sessionVariables = null;
```
Comment (howrocks) [inline fe/fe-core/src/main/java/com/starrocks/load/loadv2/LoadingTaskPlanner.java:104]: unused member？

File: fe/fe-core/src/test/java/com/starrocks/planner/OlapTableSinkTest.java [hunk 5]
```diff
@@ -123,7 +124,7 @@ public void testSinglePartition() throws UserException {
             result = partition;
         }};
 
-        OlapTableSink sink = new OlapTableSink(dstTable, tuple, Lists.newArrayList(2L));
+        OlapTableSink sink = new OlapTableSink(dstTable, tuple, Lists.newArrayList(2L), TWriteQuorumType.MAJORITY);
```
Comment (howrocks) [inline fe/fe-core/src/test/java/com/starrocks/planner/OlapTableSinkTest.java:127]: You can use pesudo cluster to add some test cases for quorum by simulating failed BE.

File: be/src/storage/segment_replicate_executor.cpp (hunk 3 above)
Comment (meegoo) [reply be/src/storage/segment_replicate_executor.cpp:197]: not necessary, primary replica should be the success one

File: fe/fe-core/src/main/java/com/starrocks/load/loadv2/LoadingTaskPlanner.java (hunk 4 above)
Comment (meegoo) [reply fe/fe-core/src/main/java/com/starrocks/load/loadv2/LoadingTaskPlanner.java:104]: yes, it will be use in another pr.

**Comment by meegoo** [comment]
run starrocks_fe_unittest

File: be/src/storage/segment_replicate_executor.cpp [hunk 6]
```diff
@@ -193,8 +193,13 @@ ReplicateToken::ReplicateToken(std::unique_ptr<ThreadPoolToken> replicate_pool_t
...
+    } else if (opt->write_quorum == WriteQuorumTypePB::ALL) {
+        _max_fail_replica_num = 0;
+    } else {
+        _max_fail_replica_num = opt->replicas.size() - (opt->replicas.size() / 2 + 1);
```
Comment (chaoyli) [inline be/src/storage/segment_replicate_executor.cpp:201]:
```suggestion
        _max_fail_replica_num = (opt->replicas.size() - 1) /2;
```

File: fe/fe-core/src/main/java/com/starrocks/persist/EditLog.java [hunk 7]
```diff
@@ -1400,6 +1401,10 @@ public void logModifyEnablePersistentIndex(ModifyTablePropertyOperationLog info)
         logEdit(OperationType.OP_MODIFY_ENABLE_PERSISTENT_INDEX, info);
     }
 
+    public void logModifyWriteQuorum(ModifyTablePropertyOperationLog info) {
+        logEdit(OperationType.OP_MODIFY_WRITE_QUORUM, info);
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/persist/EditLog.java:1405]: Does it throw a compatibility problem? @gengjun-git

File: gensrc/thrift/AgentService.thrift [hunk 8]
```diff
@@ -274,7 +274,8 @@ struct TRecoverTabletReq {
 enum TTabletMetaType {
     PARTITIONID,
     INMEMORY,
-    ENABLE_PERSISTENT_INDEX
+    ENABLE_PERSISTENT_INDEX,
+    WRITE_QUORUM
```
Comment (chaoyli) [inline gensrc/thrift/AgentService.thrift:278]: The option is for table, not for one load job?

Comment (meegoo) [reply gensrc/thrift/AgentService.thrift:278]: yes

File: be/src/storage/segment_replicate_executor.cpp (hunk 6 above)
Comment (meegoo) [reply be/src/storage/segment_replicate_executor.cpp:201]: result is same

**Comment by meegoo** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by meegoo** [comment]
run starrocks_fe_unittest

**Comment by meegoo** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/persist/EditLog.java (hunk 7 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/persist/EditLog.java:1457]: No, the lower version view ignore this log

**Review by gengjun-git — DISMISSED** [review]

**Comment by meegoo** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 76 / 116 (65.52%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|2|00.00%|[1457, 1458]|
|:large_blue_circle:|com/starrocks/catalog/ExternalOlapTable.java|0|1|00.00%|[330]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|2|17|11.76%|[2099, 2100, 3856, 3857, 3858, 3859, 3860, 3862, 3864, 3866, 3867, 3868, 3869, 3877, 3878]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|1|5|20.00%|[139, 140, 143, 144]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|4|25.00%|[2134, 2135, 2136]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|5|11|45.45%|[1105, 1220, 1221, 1222, 1223, 1224]|
|:large_blue_circle:|com/starrocks/alter/Alter.java|3|6|50.00%|[515, 524, 525]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|6|8|75.00%|[120, 121]|
|:large_blue_circle:|com/starrocks/common/util/WriteQuorum.java|11|13|84.62%|[9, 33]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|6|7|85.71%|[464]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|8|9|88.89%|[1646]|
|:large_blue_circle:|com/starrocks/transaction/OlapTableTxnStateListener.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/StreamLoadPlanner.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/DeletePlanner.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadingTaskPlanner.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/TransactionChecker.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/DatabaseTransactionMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/UpdatePlanner.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|3|3|100.00%|[]|

**Review by chaoyli — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Pull request merged**